crates/core/tedge_agent/src/software_manager/actor.rs

```diff
@@ -101,7 +104,14 @@ impl Actor for SoftwareManagerActor {
 
         while let Some(request) = input_receiver.recv().await {
             tokio::select! {
-                _ = self.handle_request(request, &mut plugins, &operation_logs) => {}
+                res = self.handle_request(request, &mut plugins, &operation_logs) => {
+                    if let Err(SoftwareManagerError::Shutdown) = res {
+                        info!("Received shutdown request while handling the operation. Exiting...");
+                        // Returning an error here instead of break to exit with a non-zero exit code
+                        // return Err(RuntimeError::Shutdown);
+                        panic!("Shutting down...")
```

Comment:
This panic is to be replaced with a proper shutdown request sent to the actor runtime, so that it shuts down all the actors cleanly, before exiting with a non-zero exit code.

Reply:
> This panic is catched by the runtime and has strictly the same effect as returning an error.

Hmm... But, that's not what I observed while testing. The agent process was never exiting when I was just returning the error. It was just stuck/idle after just shutting down the `SoftwareManagerActor`. But with the panic, the process was always exiting, triggering the restart as follows:

```
Jan 12 15:19:58 d5b8bb2b0f8a tedge-agent[490]: 2024-01-12T15:19:58.47165438Z  INFO tedge_agent::software_manager::actor: Self update detected. Requesting shutdown...
Jan 12 15:19:58 d5b8bb2b0f8a tedge-agent[490]: 2024-01-12T15:19:58.47166938Z  INFO tedge_agent::software_manager::actor: Received shutdown request while handling the operation. Exiting...
Jan 12 15:19:58 d5b8bb2b0f8a tedge-agent[490]: thread 'tokio-runtime-worker' panicked at 'Shutting down...', crates/core/tedge_agent/src/software_manager/actor.rs:112:25
Jan 12 15:19:58 d5b8bb2b0f8a tedge-agent[490]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Jan 12 15:19:58 d5b8bb2b0f8a systemd[1]: tedge-agent.service: Main process exited, code=killed, status=6/ABRT
Jan 12 15:19:58 d5b8bb2b0f8a systemd[1]: tedge-agent.service: Failed with result 'signal'.
Jan 12 15:20:03 d5b8bb2b0f8a systemd[1]: tedge-agent.service: Scheduled restart job, restart counter is at 1.
Jan 12 15:20:03 d5b8bb2b0f8a systemd[1]: Stopped tedge-agent.service - tedge-agent is a thin-edge.io component to support operations..
Jan 12 15:20:03 d5b8bb2b0f8a systemd[1]: Starting tedge-agent.service - tedge-agent is a thin-edge.io component to support operations....
Jan 12 15:20:03 d5b8bb2b0f8a systemd[1]: Started tedge-agent.service - tedge-agent is a thin-edge.io component to support operations..
```

> The runtime has to be fixed to shutdown all the actors when one fails with an error.

Thanks for fixing that.

Reply:
> Hmm... But, that's not what I observed while testing. The agent process was never exiting when I was just returning the error. It was just stuck/idle after just shutting down the `SoftwareManagerActor`.

Yes, this was the expected behavior before https://github.com/thin-edge/thin-edge.io/commit/ed8f5a33ec82f8faf0faefd82e9c271ecbc9917b

> But with the panic, the process was always exiting, triggering the restart as follows:

Here, I don't understand something. First, I observed something different (the panic stopping only the `SoftwareManagerActor`). Second, a test assesses the contrary https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/tedge_actors/src/runtime.rs#L523.

Reply:
> Here, I don't understand something.

This what I missed: [`panic = "abort"`](https://github.com/thin-edge/thin-edge.io/blob/main/Cargo.toml#L191) for `profile.release`. I was testing the debug release.

We might have to reconsider this.